go/multipooler/manager/consensus_state.go

```diff
@@ -0,0 +1,208 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manager
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	clustermetadatapb "github.com/multigres/multigres/go/pb/clustermetadata"
+	multipoolermanagerdatapb "github.com/multigres/multigres/go/pb/multipoolermanagerdata"
+)
+
+// ConsensusState manages the in-memory and on-disk consensus state for this node.
+// It provides thread-safe access to consensus state and ensures that memory is only
+// updated after successful disk writes (pessimistic approach).
+type ConsensusState struct {
+	poolerDir string
+	serviceID *clustermetadatapb.ID
+
+	mu   sync.Mutex
+	term *multipoolermanagerdatapb.ConsensusTerm // cached term from disk
+}
+
+// NewConsensusState creates a new ConsensusState manager.
+// It does not load state from disk - call Load() to initialize.
+func NewConsensusState(poolerDir string, serviceID *clustermetadatapb.ID) *ConsensusState {
+	return &ConsensusState{
+		poolerDir: poolerDir,
+		serviceID: serviceID,
+		term:      nil,
+	}
+}
+
+// Load loads consensus state from disk into memory.
+// If the file doesn't exist, initializes with default values (term 0, no accepted leader).
+// This method is idempotent - subsequent calls will reload from disk.
+func (cs *ConsensusState) Load() error {
+	term, err := getConsensusTerm(cs.poolerDir)
+	if err != nil {
+		return fmt.Errorf("failed to load consensus term: %w", err)
+	}
+
+	cs.mu.Lock()
+	cs.term = term
+	cs.mu.Unlock()
+
+	return nil
+}
+
+// GetCurrentTermNumber returns the current term.
+// Returns 0 if state has not been loaded.
+func (cs *ConsensusState) GetCurrentTermNumber() int64 {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if cs.term == nil {
+		return 0
+	}
+	return cs.term.GetTermNumber()
+}
+
+// GetAcceptedLeader returns the candidate ID this pooler accepted as leader in the current term.
+// Returns empty string if no leader was accepted.
+func (cs *ConsensusState) GetAcceptedLeader() string {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if cs.term == nil || cs.term.AcceptedLeader == nil {
+		return ""
+	}
+	return cs.term.AcceptedLeader.GetName()
+}
+
+// GetTerm returns a copy of the current consensus term.
+// Returns nil if state has not been loaded.
+func (cs *ConsensusState) GetTerm() *multipoolermanagerdatapb.ConsensusTerm {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if cs.term == nil {
+		return nil
+	}
+
+	// Return a copy to prevent external modifications
+	return cloneTerm(cs.term)
+}
+
+// PrepareAcceptance creates a new term with the acceptance recorded.
+// This does NOT modify in-memory state - the caller must call SaveAndUpdate.
+// Returns an error if already accepted a different candidate in this term.
+func (cs *ConsensusState) PrepareAcceptance(candidateID string) (*multipoolermanagerdatapb.ConsensusTerm, error) {
```

Comment:
Could we add some comments on when we will use `PrepareAcceptance` vs `PrepareTermUpdate`. In this context it's not obvious why we need this nuance.

Reply:
In the context of BeginTerm, there are two distinct phases. The first one is revocation. Updating the term with nil for AcceptedLeader and the new term_number persists the revocation decision. The next step is accepting the candidate as the new leader. That gets persisted separately.

Reply:
Right, I meant add comment to the code so future readers can have that context